consensus/state.go

```diff
@@ -2161,6 +2172,11 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
 	added, err = cs.Votes.AddVote(vote, peerID, extEnabled)
 	if !added {
 		// Either duplicate, or error upon cs.Votes.AddByIndex()
+
+		// If the vote wasnt added but there's no error, its a duplicate vote
```

Comment:
It's hard to follow the logic in this monster function. It's not a fault of this PR in particular, though.